query-planner-js/src/Scope.ts

```diff
@@ -0,0 +1,259 @@
+ * }
+ * ```
+ * still assuming that `v` is of type `V` and assuming that X is a subtype of V, the scope for field
+ * `f` will simply be `V <- X` (instead of `V <- X <- V`) as the outermost `... on V` condition adds
+ * nothing (but scope always preserve the most inner type condition as for a field, this is the type
+ * in which the field is defined). Additionally, if type `X` happens to a _super type_ of `V`, the
+ * scope for `f` would simply be `V`.
+ */
+export class Scope {
+  private cachedRuntimeTypes?: ReadonlyArray<GraphQLObjectType>;
+  private cachedIdentityKey: string | undefined = undefined;
+
+  private constructor(
+    private readonly context: QueryPlanningContext,
+
+    /** The innermost type of this scope. */
+    public readonly parentType: GraphQLCompositeType,
+    /** Directives associated to `parentType`. */
+    public readonly directives?: ReadonlyArray<DirectiveNode>,
+    /** Then enclosing scope, that is the next scope in the chain. */
+    public readonly enclosing?: Scope
+  ) {
+  }
+
+  /**
+   * Creates a new scope given a top-level type.
+   *
+   * @param context - the query planning context in which this scope is created.
+   * @param parentType - the top-level (and only) type of the created scope.
+   * @returns the newly created scope.
+   */
+  static create(context: QueryPlanningContext, parentType: GraphQLCompositeType) {
+    return new Scope(context, parentType, undefined, undefined);
+  }
+
+  /**
+   * Refines this scope by (maybe) adding the provide type (with the optional provided directives).
+   *
+   * @param type - the new type to refine this scope with.
+   * @param directives - optional directives for `type` condition.
+   * @returns a scope corresponding to restrictive this scope with the `type`. This may return
+   * this scope (be a no-op) if `directives` is undefined/null/empty and `type` is a super type of
+   * one of the existing type in this scope. Otherwise (if either `directives` is non empty or
+   * `type` genuinely restricts the existing types in this scope), a new scope with `type`
+   * as first link in the scope chain.
+   */
+  refine(type: GraphQLCompositeType, directives?: ReadonlyArray<DirectiveNode>) : Scope {
+    // Always treat the absence of directives as "undefined" to make is simpler.
+    if (directives && directives.length == 0) {
+      directives = undefined;
+    }
+    // If we have directives, we always want to preserve the condition so as to preserve that directive.
+    if (directives) {
+      return new Scope(this.context, type, directives, this);
+    }
+    // Scope always preserve the immediate parent type, but if the new type is already the parent one
+    // and we have not directives to preserve, simply return the existing scope.
+    if (!directives && type === this.parentType) {
+      return this;
+    }
+    return new Scope(this.context, type, directives, Scope.pruneRefinedTypes(this, type));
+  }
+
+  private static pruneRefinedTypes(
```

Comment:
Clearly, I bring some habits from Java. That said, I do slightly prefer a `private static` in this case just because `pruneRefinedTypes` is very just an helper for `refine` and this allow to put both method physically close by in the file.

That said, if you think `private static` is sufficiently uncommon in typescript that its worth avoiding, I could easily turn this into a proper (private) method (it's static to slightly simplify the handling of `undefined`, but I could handle it at the call sites instead).

Reply:
I don't think making it non-static is necessarily better, if there are no dependencies on the instance. I think a top-level function (private to the module) would be more common, but I'm fine with leaving this as `private static` and perhaps following that example in other places if we think it makes sense there as well.